### PR TITLE
Test cases to verify port shaping rate and priority scheduling 

### PR DESCRIPTION
**Pull request by ersuneetsingh**
In this PR, the FLiRT test case is added to verify the port shaping rate. Also, the chassis config is added to support shaping rate in stratum.

### CODE REVIEW

**Review by ccascone — COMMENTED** [review]
Thanks @ersuneetsingh. Overall looks good, I left mostly minor comments to clean-up improve usability.

**Comment by ccascone** [comment]
@ersuneetsingh please check the log of the failed checks. Those issues will need to e fixed before merging.

File: trex-scripts/tests/port_shaping_rate.py
Comment (ersuneetsingh) [reply trex-scripts/tests/port_shaping_rate.py:34]: Yes I agree. We can put "mult=100%" in line 54, which basically means to send traffic on line rate (i.e., 100% bandwidth utilization ). For other option without inserting "mult" value, I need to check.

File: trex-scripts/tests/port_shaping_rate.py
Comment (ersuneetsingh) [reply trex-scripts/tests/port_shaping_rate.py:36]: Yes, -1 indicates 'unlimited'. Ok I agree to change the default to 10.

File: trex-scripts/tests/port_shaping_rate.py
Comment (ersuneetsingh) [reply trex-scripts/tests/port_shaping_rate.py:45]: Sure. Thanks.

File: trex-scripts/tests/port_shaping_rate.py
Comment (ersuneetsingh) [reply trex-scripts/tests/port_shaping_rate.py:70]: Thanks for this suggestion. I will add  the assertion statement.

File: trex-scripts/tests/port_shaping_rate.py
Comment (Yi-Tseng) [reply trex-scripts/tests/port_shaping_rate.py:34]: More detail about the "mult" of STLClient API can be found here:

```
mult : str
                    Multiplier in a form of pps, bps, or line util in %
                    Examples: "5kpps", "10gbps", "85%", "32mbps"
```

https://github.com/cisco-system-traffic-generator/trex-core/blob/master/scripts/automation/trex_control_plane/interactive/trex/stl/trex_stl_client.py#L684-L686

File: trex-scripts/tests/priority_scheduling.py
Comment (ersuneetsingh) [inline trex-scripts/tests/priority_scheduling.py:24]: class name needs to be change to "PriorityScheduling".

File: trex-scripts/tests/port_shaping_rate.py
Comment (ccascone) [inline trex-scripts/tests/port_shaping_rate.py:22]: The convention for constants is to use upper case. Also adding the unit to the name usually improves readability.
```suggestion
SHAPING_RATE_MBPS  = 1000 # in Mbps
```

File: trex-scripts/tests/port_shaping_rate.py
Comment (ccascone) [inline trex-scripts/tests/port_shaping_rate.py:59]: Same comment here about adding unit to the variable name
```suggestion
        rx_rate_mbps = stats[1]['rx_bps'] / (10**6)
```

**Review by ccascone — APPROVED** [review]

File: trex-scripts/tests/priority_scheduling.py
Comment (ccascone) [inline trex-scripts/tests/priority_scheduling.py:110]: As discussed yesterday, would be good to add an assertion that checks the maximum latency for the delay-critical flow.

File: trex-scripts/tests/priority_scheduling.py
Comment (ccascone) [inline trex-scripts/tests/priority_scheduling.py:24]: Ideally, this test should automate testing with and without priorities, but I guess before we can do that we need to wait for @Yi-Tseng to complete the integration with the p4runtime library that is happening in fabric-tna. So maybe we can do that in a later PR.

File: trex-scripts/tests/priority_scheduling.py
Comment (ersuneetsingh) [reply trex-scripts/tests/priority_scheduling.py:110]: Yes sure, can we go with upto 100 microseconds for delay critical flow ?

File: trex-scripts/tests/priority_scheduling.py
Comment (ersuneetsingh) [reply trex-scripts/tests/priority_scheduling.py:24]: Nice. Got it.

File: stratum/chassis_config.pb_with_PortShaping.txt
Comment (Yi-Tseng) [inline stratum/chassis_config.pb_with_PortShaping.txt:15]: fix the indent

**Review by Yi-Tseng — COMMENTED** [review]
lgtm, please resolve comments from Carmelo

File: trex-scripts/tests/port_shaping_rate.py
Comment (ccascone) [reply trex-scripts/tests/port_shaping_rate.py:43]: Tee line is still here :)

File: trex-scripts/tests/priority_scheduling.py
Comment (ccascone) [reply trex-scripts/tests/priority_scheduling.py:117]: Maybe, I'll leave that to your judgment. You have more experience than me running these tests and a better feeling of what is the error/overhead introduced by trex.

If you want to go one extra step further and be extra sure that prioritization is making a difference, I think the assertion could check the difference with and without priority. To do that, maybe you can have two delay-critical flows (so three flows in total), one that is prioritized and one that is not, and compare the latency in the assertion? For example: `assert max_latency_without_qos - max_latency_with_qos > SOME_KNOWN CONSTANT`.

File: stratum/chassis_config.pb_with_PortShaping.txt
Comment (Yi-Tseng) [inline stratum/chassis_config.pb_with_PortShaping.txt:31]:
```suggestion
vendor_config {
  tofino_config {
    node_id_to_port_shaping_config {
      key: 1
      value {
        per_port_shaping_configs {
          key:30
          value {
            byte_shaping {
              max_rate_bps: 1000000000 # 1Gbps
              max_burst_bytes: 160000 # 2x jumbo frame
            }
          }
        }
      }
    }
  }
}
```

File: trex-scripts/tests/priority_scheduling.py
Comment (ccascone) [inline trex-scripts/tests/priority_scheduling.py:25]: nit: the convention for constants is all caps
```suggestion
LATENCY_DC_MAX_USEC = 1000   # in microseconds
```

File: trex-scripts/tests/priority_scheduling.py
Comment (ccascone) [inline trex-scripts/tests/priority_scheduling.py:28]:
```suggestion
LATENCY_LP_MAX_USEC = 24000 # in microseconds
```

File: trex-scripts/tests/priority_scheduling.py
Comment (ccascone) [inline trex-scripts/tests/priority_scheduling.py:57]: It would make the code more readable if you could set `pg_id` in a consistent way following the stream variable names, i.e. 1, 2, 3. So `s1` should have `pg_id = 1`, `s2` should have `pg_id = 2`, etc.

File: trex-scripts/tests/priority_scheduling.py
Comment (ccascone) [inline trex-scripts/tests/priority_scheduling.py:89]: Similar to the previous comment, you could call this `flow_stats_1` so it is obvious that these are the stats associated with `s1`, otherwise it could be confusing to try to remember what pg_id 5 was used for. The same comment applies to other variables defined below which have the pg_id in their name.
```suggestion
        flow_stats_1       = stats['flow_stats'].get(1)
```

File: trex-scripts/tests/priority_scheduling.py
Comment (ccascone) [inline trex-scripts/tests/priority_scheduling.py:103]: Same here, let's use stream numbers in printed messages
```suggestion
        print(" \n TX and RX flow stats and packets dopped for s1 (i.e., delay critical): ")
```

File: trex-scripts/tests/priority_scheduling.py
Comment (ccascone) [inline trex-scripts/tests/priority_scheduling.py:109]:
```suggestion
        print(" \n TX and RX flow stats and packet dopped for s2 (i.e., delay critical): ")
```

**Review by ccascone — COMMENTED** [review]
I left only minor comments. I think we should merge after you fix them.

**Comment by ersuneetsingh** [comment]
> I left only minor comments. I think we should merge after you fix them.

Thanks for all your suggestions. I have already fixed the required changes.

**Comment by ccascone** [comment]
@Yi-Tseng ok to merge?

File: stratum/chassis_config.pb_with_PortShaping.txt
Comment (pudelkoM) [inline stratum/chassis_config.pb_with_PortShaping.txt:2]: copy&paste error?

File: stratum/chassis_config.pb_with_PortShaping.txt.license
Comment (pudelkoM) [inline stratum/chassis_config.pb_with_PortShaping.txt.license:2]: Add as comment directly in file (`# comment`)

File: stratum/chassis_config.pb_with_PortShaping.txt
Comment (ersuneetsingh) [reply stratum/chassis_config.pb_with_PortShaping.txt]: Thanks for pointing this error. I already resolved.

File: stratum/chassis_config.pb_with_PortShaping.txt.license
Comment (ersuneetsingh) [reply stratum/chassis_config.pb_with_PortShaping.txt.license]: thanks. I already resolved.

File: stratum/chassis_config.pb_with_PortShaping.txt
Comment (Yi-Tseng) [inline stratum/chassis_config.pb_with_PortShaping.txt:1]: rename to `chassis_config_with_port_shaping.pb.txt`

**Review by Yi-Tseng — APPROVED** [review]
lgtm overall, just one comment

File: stratum/chassis_config.pb_with_PortShaping.txt
Comment (ersuneetsingh) [reply stratum/chassis_config.pb_with_PortShaping.txt]: thanks for your suggestion. Done.

Comment (Yi-Tseng) [reply stratum/chassis_config.pb_with_PortShaping.txt]: still incorrect, can you update it again?

Comment (ccascone) [reply stratum/chassis_config.pb_with_PortShaping.txt]: It looks fixed to me

Comment (ersuneetsingh) [reply stratum/chassis_config.pb_with_PortShaping.txt]: > still incorrect, can you update it again?

yes , it was not correctly updated. I already fixed the file name now.

**Pull request merged**
